### ISSUE TITLE
Regression - Hardware Deinterlacing not used with MPEG2 content on older ATI Radeon HD 6450 Video Card

### ISSUE THREAD

**Issue by therealjoeblow** [xbmc/xbmc#24530]
### Describe the bug
Vector Adaptive hardware deinterlacing is not being used when playing back interlaced MPEG2 content on v21. 

My older ATI Radeon HD 6450 supports vector adaptive hardware deinterlacing.  On Kodi Nexus v20 and previous this was working fine, the interlaced MPEG2 remuxed DVD content that I have was being deinterlaced on the hardware properly and the resulting playback looked basically progressive with no interlacing artifacts.  On Omega v21 now the deinterlacing is not being engaged and the content has the "window-blind" or "stair-step" interlacing effect still intact on playback.

I realize this is an older card, and there are no new drivers available for many years now.  However the card has worked fine for standard H264 and MPEG playback till now.  I don't believe it is specifically a driver issue as the hardware deinterlacing was working as expected on v20 and earlier versions.  I believe it is a regression with the switch to FFMPEG v6 in Kodi v21 that is not triggering the hardware deinterlacing as expected.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

I would expect the hardware deinterlacing to be engaged and the content playback to be smooth and artifact free

## Actual Behavior

The content plays back without being deinterlaced

## Possible Fix

I don't know.

### To Reproduce

1. Play interlaced MPEG2 content (a standard DVD file remuxed into a .MKV container)
2.
3.

### Debuglog

The debuglog can be found here:

https://paste.kodi.tv/jusovawari.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [X] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10 22H2 x64
 - Kodi version: 21.0-BETA2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by therealjoeblow** [comment]
Here's a log file of v20 playing back the same file:

https://paste.kodi.tv/boliyisecu.kodi

And here are a couple of comparative screenshots:

v21:
https://i.postimg.cc/Qt6xNJ2j/Screenshot-2024-01-17-225655.png

v20:
https://i.postimg.cc/wM7x1szL/Screenshot-2024-01-17-230312.png

**Comment by CrystalP** [comment]
That gpu/driver is an oddball (also it's a 7400 IGP according to the log, not a 6450)

```
DXVA::CEnumeratorHD::ProbeProcessorCaps: selected video processor recommends 0 future frames and 0 past frames for best results.
```

Can you share a short sample of that video, there are a couple other strange things happening, with a weird initial refresh rate of 12.5

The deinterlacing problems likely have to do with some changes I made in the dxva renderer for NVIDIA non-standard behavior, ffmpeg itself is probably fine.

**Comment by CrystalP** [comment]
If you really have a 6450 is it possible that somehow you managed to install the wrong driver?

**Comment by therealjoeblow** [comment]
I will make you a sample clip of the video file tonight after work.

**Comment by therealjoeblow** [comment]
Here is a clip of that MPEG2 file, let me know if you can download it (MediaFire):

https://www.mediafire.com/file/nr53o3twyw3dqe7/VIDEO.PAL.DVD9.MPEG2-CLIP.mkv/file

**Comment by CrystalP** [comment]
Thanks I was able to download it. Is it bad for you for the entire clip or only after a few seconds?

I see bad quality deinterlacing a few seconds in on recent AMD, but it's not combing or a lack of deinterlacing, and looks a bit like a lower vertical resolution (likely half-resolution).
NVIDIA is perfect on the same but has a higher quality algorithm to begin with and some "non-standard magic", so not directly comparable.
Still I'll dig a bit more, see if something can be done. Could be related to another change I did to start a stream immediately in the correct fps, or it was pre-existing,

**Comment by therealjoeblow** [comment]
It does that for the whole video, not just parts, whether I start from the beginning and just let it play or if I skip to the middle, or resume start next time from where I left off.  Consistently not deinterlacing.  And all the videos in the series from Season 1, 2, 3 and 4, which DVD's were all released at different times and some are from R2 and some from R4 that I bought on eBay so not an issue with the DVD or it's encoding as they are all from different sources.  And many other remuxed DVD's I have in my collection are interlaced and do the exact same thing.

I also have some other interlaced test clips that I routinely use to test for proper deinterlacing, both PAL and NTSC versions.  I posted them here for you to check as well.  It does the same with these, no deinterlacing applied, bad combing throughout.

https://www.mediafire.com/file/h3y05idfjw190v0/SD_NTSC_29.97_640x480.ts/file
https://www.mediafire.com/file/bcb4i49p4bh5mck/SD_PAL_25_576i.ts/file

All of these play using the GPU and proper deinterlacing and look progressive outside of Kodi with MPC, I just rechecked to be 100% certain, and as I said above, all of them were being deinterlaced in v20 and before just fine without any issue at all.  Just now none of them are working anymore in v21.

**Comment by therealjoeblow** [comment]
Just thinking out loud here as I really don't know that much about coding...

..On your earlier comment: _"The deinterlacing problems likely have to do with some changes I made in the dxva renderer for NVIDIA non-standard behavior"_

Could it be as simple as to keep the old code there in a subloop and add  some logic check like "If AMD->[use old code]", and "If NVIDIA->[use new code]"?  Since the old code worked fine on AMD?  Or is it not that simple?

**Comment by CrystalP** [comment]
if only :-) I have a pretty good idea of the area to look at, but there may be some trial and error since I don't see the problem on my computer and won't know if a change fixes the problem until you test and confirm.
From what I gather you don't build your own Kodi, I'd have to provide some test builds?

**Comment by therealjoeblow** [comment]
That's correct, I don't know how to build Kodi mysel from the source.  I am currently using a v21 beta build that thexai made for me to fix an audio regression, it's based on the nightly code from 3 days ago.

**Comment by therealjoeblow** [comment]
Another thing I just noticed, might help you track it down, not sure.  When playing back any kind of MPEG2 content, almost exactly 5 seconds after playback starts there is a full screen black flash, lasts a fraction of a second then the picture resumes normally (well, with the combing if it's interlaced).  Happens every time you play anything with MPEG2.  It does NOT happen at all when playing H264, XviD, H265 or VC-1.  ^only^ with MPEG2.

**Comment by CrystalP** [comment]
Please try these 5 builds and report the results for interlaced and progressive video please. They each have a slightly different way to attempt to fix the issue.

http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20240125-c9fd66be--x64.exe
http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20240125-ef18fd25--x64.exe
http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20240125-26d628c8--x64.exe
http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20240125-db256c33--x64.exe
http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20240125-18d1057d--x64.exe

**Comment by therealjoeblow** [comment]
I tested all 5 of those @CrystalP , unfortunately none of them fix the problem.  Deinterlacing not being applied in any of them, end result looks exactly the same as with the initial problem report, combing throughout.  Also, all 5 of them exhibit the same very brief full-screen black flash a few seconds after MPEG2 playback starts, exactly same as reported above.

**Comment by CrystalP** [comment]
That's surprising. Please try the previous releases of v21 to get an idea of when this was broken (v21 alpha1 alpha2 alpha3 beta1 beta2)

When you wrote that this was working for v20, did you mean the latest 20.3?

**Comment by therealjoeblow** [comment]
@CrystalP could you please point me to the installers for alpha1/2/3 so I can test those?  I can only find the Beta 1 and 2 in the release library.  Thx.

**Comment by fuzzard** [comment]
Pre beta releases can be found at https://mirrors.kodi.tv/snapshots/windows/win64/

**Comment by therealjoeblow** [comment]
Test results as follows:

Alpha 1 and 2 were fine, deinterlacing was being properly applied as in v20.x

Alpha 3 broke it, no longer deinterlacing, and on top of that *everything* (interlaced mpeg2 and progressive h264) all play back with a blue/purple hue over them:

[https://i.postimg.cc/3N0DzrDp/Screenshot-2024-02-05-185116.png](url)
[https://i.postimg.cc/t70Vb0hc/Screenshot-2024-02-05-185208.png](url)

That stayed the same in Beta 1 and Beta 2 (broken deinterlacing, plus blue tinge)

In Beta 3 the deinterlacing is still broken, but the blue tinge is gone and now the colors are correct, as per the screenshots with the broken deinterlacing in my original problem report.

**Comment by therealjoeblow** [comment]
Deinterlaces properly in the first two, broken in the third one (88e8928).

**Comment by CrystalP** [comment]
Please test the following:
4dcaf8634c8519fb946557704831a991205955f1 https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20230702-4dcaf863--x64.exe
1c65bbbe61e3dfb4b832d3baa431001551bf7520 https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20230709-1c65bbbe--x64.exe
60ae8f5659f46a4295acb9711f9394b420d0d3be https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20230718-60ae8f56--x64.exe

**Comment by therealjoeblow** [comment]
@CrystalP , my apology for not getting back to you sooner.

Deinterlacing does NOT work in 4dcaf86
It DOES work in both of the others (1c65bbb and 60ae8f5)

**Comment by CrystalP** [comment]
To sum up, it was broken once by https://github.com/xbmc/xbmc/commit/4dcaf8634c8519fb946557704831a991205955f1, fixed by https://github.com/xbmc/xbmc/commit/1c65bbbe61e3dfb4b832d3baa431001551bf7520, re-broken by https://github.com/xbmc/xbmc/commit/88e89285bffa72a7da3394e257d800be2185f8f2 and has been broken since then.

They are respectively https://github.com/xbmc/xbmc/pull/23463, https://github.com/xbmc/xbmc/pull/23484 and https://github.com/xbmc/xbmc/pull/23552

@thexai do you have an idea for https://github.com/xbmc/xbmc/pull/23552? It was not supposed to change anything.

@therealjoeblow could you please turn on component-specific logging and the component category "Video" and provide debug logs of playing the clip with the https://github.com/xbmc/xbmc/commit/60ae8f5659f46a4295acb9711f9394b420d0d3be and https://github.com/xbmc/xbmc/commit/88e89285bffa72a7da3394e257d800be2185f8f2 versions.

**Comment by CrystalP** [comment]
Yes I'm aware that new AMD RDNA+ hw and drivers started having issues and it's worrying but without the hw on hand, what can we do... As far as I know those crashes happen for any Kodi version (any v21, v20 v19).

This one is different: the gpu or driver didn't change and they're both old/legacy, and the Kodi version makes the difference between good deinterlacing and no deinterlacing. Something committed between 21alpha2 and 21alpha3 caused it.

I tried extracting the relevant PR for bisection. It's something after https://github.com/xbmc/xbmc/commit/60ae8f5659f46a4295acb9711f9394b420d0d3be (known good) until (and including) https://github.com/xbmc/xbmc/commit/88e89285bffa72a7da3394e257d800be2185f8f2 (known bad)
I'll look for another possible culprit in the range just in case I missed something.

**Comment by therealjoeblow** [comment]
@thexai that's for newer models with current drivers.  I'm talking about an old legacy card here which has had perfectly working vector adaptive deinterlacing in MPEG2 video for a decade now.  And it worked perfectly with Kodi v19 and v20, and outside of v21 it works perfectly with MPC-HC as well with the old legacy drivers.  The last driver for this card was released *ages* ago and is not broken.

@CrystalIP I'll test those builds with the logging in the next day or 2.

**Comment by therealjoeblow** [comment]
Hi @CrystalP sorry for the long hiatus.  I gave up on the Old ATI/AMD card and bought a GeForce GT1030.  It just works so much better, there are too many glitches and bad coding with the AMD drivers.  Been using various incarnations of ATI/AMD cards forever (since back in the mid 1990's, wow, around 30 years!) but finally broke the mold and moved on to the otherside. I always thought... better the devil you know, but in hindsight, ATI/AMD cards have always been a struggle to get everything working right, and then just when you do, a new driver comes out and breaks something else.

Sorry for the dead end here, I do very much appreciate the effort you put into trying to solve this.  I guess you can just close the issue and mark it unresolved.

Full high quality deinterlacing and all other aspects of hardware decoding (including H265 and HDR tonemapping) work perfectly on the GT1030 (fanless, no noise).

Thanks again for the help though.
